File: 1091-ShortestPathinBinaryMatrix/1091-ShortestPathinBinaryMatrix.java
/*

1091. Shortest Path in Binary Matrix
Medium

In an N by N square grid, each cell is either empty (0) or blocked (1).

A clear path from top-left to bottom-right has length k if and only if it is composed of cells C_1, C_2, ..., C_k such that:

    Adjacent cells C_i and C_{i+1} are connected 8-directionally (ie., they are different and share an edge or corner)
    C_1 is at location (0, 0) (ie. has value grid[0][0])
    C_k is at location (N-1, N-1) (ie. has value grid[N-1][N-1])
    If C_i is located at (r, c), then grid[r][c] is empty (ie. grid[r][c] == 0).

Return the length of the shortest such clear path from top-left to bottom-right.  If such a path does not exist, return -1.

Example 1:

Input: [[0,1],[1,0]]


Output: 2

Example 2:

Input: [[0,0,0],[1,1,0],[1,1,0]]

Output: 4
 

Note:

    1 <= grid.length == grid[0].length <= 100
    grid[r][c] is 0 or 1

*/

class Solution {
    public int shortestPathBinaryMatrix(int[][] grid) {
        Queue<int[]> q = new LinkedList<>();
        if(grid[0][0] == 1) return -1;
        if(grid[grid.length - 1][grid[0].length - 1] == 1) return -1;
        q.add(new int[]{0, 0});
        int count = 1;
        int[][] dirs = new int[][]{{0, 0},{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}};
        while(!q.isEmpty()){
            count++;
            int len = q.size();
            for(int i = 0; i < len; i++){
                int[] temp = q.poll();
                for(int[] dir : dirs){
                    int x = temp[0] + dir[0];
                    int y = temp[1] + dir[1];
                    if(x < 0 || x >= grid.length || y < 0 || y >= grid[0].length || grid[x][y] == 1) continue;
                    if(x == grid.length - 1 && y == grid[0].length - 1) return grid.length == 1 ? count - 1 : count;
                    grid[x][y] = 1;
                    q.add(new int[]{x, y});
                }
            }
        }
        return -1;
    }
}

File: 1091-ShortestPathinBinaryMatrix/1091-ShortestPathinBinaryMatrix.py
"""

1091. Shortest Path in Binary Matrix

Given an n x n binary matrix grid, return the length of the shortest clear path in the matrix. If there is no clear path, return -1.

A clear path in a binary matrix is a path from the top-left cell (i.e., (0, 0)) to the bottom-right cell (i.e., (n - 1, n - 1)) such that:

    All the visited cells of the path are 0.
    All the adjacent cells of the path are 8-directionally connected (i.e., they are different and they share an edge or a corner).

The length of a clear path is the number of visited cells of this path.

Example 1:

Input: grid = [[0,1],[1,0]]
Output: 2

Example 2:

Input: grid = [[0,0,0],[1,1,0],[1,1,0]]
Output: 4

Example 3:

Input: grid = [[1,0,0],[1,1,0],[1,1,0]]
Output: -1

Constraints:

    n == grid.length
    n == grid[i].length
    1 <= n <= 100
    grid[i][j] is 0 or 1

"""

class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        q = deque()
        if grid[0][0] == 1: return -1
        q.append((0, 0))
        vis = [[False] * len(grid) for _ in range(len(grid[0]))]
        dirs = [[0, 1], [0, -1], [1, 0], [-1, 0], [1, 1], [-1, 1], [1, -1], [-1, -1]]
        res = 0
        while q:
            length = len(q)
            res += 1
            for i in range(length):
                curr = q.popleft()
                if curr[0] == len(grid) - 1 and curr[1] == len(grid[0]) - 1: return res
                for d in dirs:
                    x = d[0] + curr[0]
                    y = d[1] + curr[1]
                    if x < 0 or y < 0 or x >= len(grid) or y >= len(grid[0]) or grid[x][y] == 1 or vis[x][y]: continue
                    vis[x][y] = True
                    q.append((x, y))
        return -1
